pkg/flowaggregator/flowaggregator.go

```diff
@@ -213,40 +234,125 @@ type flowAggregator struct {
 	sendJSONRecord              bool
 	numRecordsExported          int64
 	numRecordsReceived          int64
+	updateCh                    chan updateMsg
+	configFile                  string
+	configWatcher               *fsnotify.Watcher
+	configData                  []byte
+	APIServer                   flowaggregatorconfig.APIServerConfig
 }
 
 func NewFlowAggregator(
-	externalFlowCollectorAddr string,
-	externalFlowCollectorProto string,
-	activeFlowRecTimeout time.Duration,
-	inactiveFlowRecTimeout time.Duration,
-	aggregatorTransportProtocol AggregatorTransportProtocol,
-	flowAggregatorAddress string,
-	includePodLabels bool,
 	k8sClient kubernetes.Interface,
-	observationDomainID uint32,
 	podInformer coreinformers.PodInformer,
-	sendJSONRecord bool,
-) *flowAggregator {
+	configFile string,
+) (*flowAggregator, error) {
 	registry := ipfix.NewIPFIXRegistry()
 	registry.LoadRegistry()
+
+	opts := util.NewOptions()
+	if err := opts.Complete(configFile); err != nil {
+		return nil, fmt.Errorf("failed to complete args: %v", err)
+	}
+	if err := opts.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate args: %v", err)
+	}
```

Comment:
IMO, there should be more symmetry in how we load the config the first time, and subsequently when the mounted file is updated. It is also a good idea to call the validation function when the config is updated, not just the first time.

We need a function like this one:
```go
func loadConfig(configBytes []byte) (*flowaggregatorconfig.FlowAggregatorConfig, error) {
    // unmarshal
    // validate
    // set defaults
}
```
that we can call in both places.

Reply:
Make sense. I've deleted most of code in option.go.
Currently the structure will be something like below:
```
data, err := ioutil.ReadFile(fa.configFile)
if err != nil {
    return fmt.Errorf("cannot read FlowAggregator configFile: %v", err)
}
config, err := util.LoadConfig(data)
if err != nil {
    return err
}
opt, err := util.Validate(config)
if err != nil {
    klog.Error(err)
    return nil
}
```

I used all of these three functions when initializing the fa pod and when we detect the changes of mounted configMap.
The only difference will be the way we handle error for Validate.
When initializing the pod, we cannot tolerate the error, so I return error in that case.
But for the updated mounted configMap, if the error happens, I just log the error and return nil. (user can still use antctl to enter the right parameters)

Hope the structure is clear now.